python/sbp/jit/msg.py

```diff
@@ -0,0 +1,323 @@
+
+
+@nb.jit('Tuple((f8,u4,u4))(u1[:],u4,u4)', nopython=True, nogil=True)
+def get_f64(buf, offset, length):
+    if length < 8:
+        return (0, offset, length)
+    res = _get_f64(buf[offset + 0],
+                   buf[offset + 1],
+                   buf[offset + 2],
+                   buf[offset + 3],
+                   buf[offset + 4],
+                   buf[offset + 5],
+                   buf[offset + 6],
+                   buf[offset + 7])
+    return res, offset + 8, length - 8
+
+
+@nb.jit('Tuple((u1[:],u4,u4))(u1[:],u4,u4,b1)', nopython=True, nogil=True)
+def _get_string(buf_in, offset, length, check_null):
+    buf_out = np.zeros(256, dtype=np.uint8)
+    i = nb.u4(0)
+    null_term = False
+    while i < length:
+        if check_null and buf_in[offset + i] == 0:
+            null_term = True
+            break
+        buf_out[i] = buf_in[offset + i]
+        i = nb.u4(i + nb.u4(1))
+    if null_term:
+        return buf_out[:i], offset + i + 1, i + 1
+    else:
+        return buf_out[:i], offset + i, i
+
+
+def get_string(buf, offset, length):
+    buf, offset, length = _get_string(buf, offset, length, True)
+    return buf.tobytes().decode('ascii'), offset, length
+
+
+def get_fixed_string(size):
+    def func(buf, offset_in, length):
+        if length < size:
+            return '', offset_in, length
+        buf, offset, length = _get_string(buf, offset_in, size, False)
+        return buf.tobytes().decode('ascii'), offset_in + size, length
+    return func
+
+
+def get_setting(buf, offset, length):
+    buf, offset, length = _get_string(buf, offset, length, False)
+    return buf.tobytes().decode('ascii'), offset, length
+
+
+def get_array(getter):
+    def func(buf, offset, length):
+        # TODO verify this function
+        arr = []
+        while length > 0:
+            o_1, (res, offset, length) = offset, getter(buf, offset, length)
+            if o_1 == offset:
+                return arr, offset, length
+            arr.append(res)
+        return arr, offset, length
+    return func
+
+
+def get_fixed_array(getter, count, el_size, nb_type=None):
+    def func(buf, offset, length):
+        offset_start = offset
+        total_size = count * el_size
+        if length < total_size:
+            return [], offset_start, length
+        arr = []
+        while total_size > 0:
+            o_1, (res, offset, length) = offset, getter(buf, offset, length)
+            if o_1 == offset:
+                return [], offset_start, length
+            arr.append(nb_type(res) if nb_type else res)
+            total_size -= el_size
+        return arr, offset, length
+    return func
+
+
+class SBP(object):
+    """Swift Binary Protocol container.  """
+
+    # Class variable for controlling whether floating point variables should be
+    # stored in numba.float32/64 format or native Python float. This affects
+    # the rounding and string representation.
+    float_meta = False
```

Comment:
@silverjam Do you think this kind of approach is OK?

Reply:
Example usage:

```Python
                    msg.SBP.float_meta = True
                    sys.stdout.write(json.dumps(res,
                                                allow_nan=False,
                                                sort_keys=True,
                                                separators=(',', ':'),
                                                cls=SbpJSONEncoder))
```

Reply:
I think it makes sense, but it exposes numba pretty directly in the message interface.   I'm not sure we can use Numba types directly if we eventually want to remove Numba as a runtime dependency, but we can cross that bridge when we come to it.

A few questions:

- What's the rationale behind the `float_meta` name?  Based on the docs, I would think something like `float_native` would be more obvious.  In either case, let's add docs for it to the SBP doc string.

- I'm assuming that `sbp2json.py` sets `float_meta` in order to be more compatible with Haskell?  Would setting this to `True` by default be more/less accurate?  My assumption was that converting from a "native" float format (regardless of size) to Python's float format wouldn't loose any information.

Reply:
Idea behind `float_meta` was to hint that it preserves the meta information whether the byte data is/was 32 or 64-bit. I think `float_native` or `float_numba` is better.

Your assumption is correct. Defaulting to `True` would produce more correct result in case of 32 bit floats as there's no fabricated decimal digits from native (64-bit) float representation. Problem with defaulting to `True` is as you already mentioned it exposes `numba.f4` and `numba.f8` types to the client and thus requires explicit support especially for `numba.f4`, see for example:

```
    def default(self, obj):
        if isinstance(obj, np.float32):
            d = dec.Decimal(np.format_float_positional(obj, precision=None, unique=True, trim='0'))
            ret = float(round(dec.Decimal(float(obj)), abs(d.as_tuple().exponent)))
            return ret
        # Let the base class default method raise the TypeError
        return json.JSONEncoder.default(self, obj)
```

`numba.f8` is at least some cases implicitly regarded as native Python float as they both are 64-bits. One possible approach to eliminate `numba` from the interface in all modes could be using the native `decimal` module, which is already utilized for imitating Haskell's rounding strategy.

Reply:
Also I am pretty positive that I am taking too many steps handling through the floats, so optimizing to be done. Current goal has been just getting out 1:1 output with Haskell with some level of sanity left in the code.

Reply:
Actually looking at the code snippet above, I remember wondering couldn't that be done already here `libsbp` side as it does return native float after the correct rounding method has been used.

Here with comments:
```
        if isinstance(obj, np.float32):
            # Let numpy's judicious rounding tell us the amount of digits we
            # want as it seems to align with Haskell's output
            d = dec.Decimal(np.format_float_positional(obj, precision=None, unique=True, trim='0'))
            # Round it using correct rounding strategy and return it as native float
            # NOTE: Can't this be done already on libsbp side?
            ret = float(round(dec.Decimal(float(obj)), abs(d.as_tuple().exponent)))
            return ret
```

Reply:
And (FYI) the rounding strategy matching with Haskell:

`dec.getcontext().rounding = dec.ROUND_HALF_UP`